### PR TITLE
chore: support replay with new file format

### PR DESCRIPTION
**Pull request by cisse21**
# Description

Support the legacy replay with newer file structure

## Linear Ticket

[ Linear Link ](https://linear.app/rudderstack/issue/PIPE-191/explore-deprecation-of-backup-mechanism)

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: app/apphandlers/embeddedAppHandler.go [hunk 1]
```diff
@@ -334,8 +334,15 @@ func (a *embeddedApp) StartRudderCore(ctx context.Context, options *app.Options)
...
-		defer replayDB.TearDown()
-		a.app.Features().Replay.Setup(ctx, &replayDB, gatewayDB, routerDB, batchRouterDB)
+		replay, err := a.app.Features().Replay.Setup(ctx, config, &replayDB, gatewayDB, routerDB, batchRouterDB)
+		if err != nil {
+			return err
+		}
+		err = replay.Start()
+		if err != nil {
+			return fmt.Errorf("could not start replay: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (cisse21) [reply app/apphandlers/embeddedAppHandler.go]: Made the changes

File: enterprise/replay/setup.go [hunk 2]
```diff
@@ -40,19 +36,46 @@ func initFileManager(log logger.Logger) (filemanager.FileManager, string, error)
...
+type Replay struct {
+	toDB        *jobsdb.Handle
+	dumpsLoader dumpsloader.DumpsLoader
+	replayer    replayer.Replay
+}
+
+type ReplayFactory interface {
+	Start() error
+	Stop() error
+}
```
[the remark this answers is not in the document]
Comment (cisse21) [reply enterprise/replay/setup.go]: Made the changes

File: enterprise/replay/dumpsloader/dumpsloader.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	startTime     time.Time
+	endTime       time.Time
+}
+
+// procErrorRequestHandler is an empty struct to capture Proc Error re-stream request handling functionality
+type procErrorRequestHandler struct {
+	g           errgroup.Group
+	tablePrefix string
+	handle      *dumpsLoaderHandle
+}
```
[the remark this answers is not in the document]
Comment (cisse21) [reply enterprise/replay/dumpsloader/dumpsloader.go]: Made the changes

File: enterprise/replay/dumpsloader/dumpsloader.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+// gwReplayRequestHandler is an empty struct to capture Gateway replay handling functionality
+type gwReplayRequestHandler struct {
+	g           errgroup.Group
+	tablePrefix string
+	handle      *dumpsLoaderHandle
+}
```
[the remark this answers is not in the document]
Comment (cisse21) [reply enterprise/replay/dumpsloader/dumpsloader.go]: Made the changes

File: enterprise/replay/dumpsloader/dumpsloader.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/rudderlabs/rudder-server/jobsdb"
+)
+
+type DumpsLoader interface {
+	Start()
+	Stop() error
+}
+
+// DumpsLoaderHandleT - dumps-loader handle
+type dumpsLoaderHandle struct {
```
[the remark this answers is not in the document]
Comment (cisse21) [reply enterprise/replay/dumpsloader/dumpsloader.go]: Made the changes

File: enterprise/replay/replayer/transformer.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		EventPayload: Payload{
+			Batch: originalPayload.Payload,
+		},
+		UUID: uuid.Must(uuid.NewRandom()).String(),
+		Parameters: Params{
+			SourceID: sourceId,
+		},
+		EventCount: 1,
+		UserID:     originalPayload.UserID,
+	}
```
Comment (atzoum) [inline enterprise/replay/replayer/transformer.go:65]:
```suggestion
	desiredPayload := struct {
		CreatedAt    time.Time `json:"created_at"`
		CustomVal    string    `json:"custom_val"`
		EventCount   int       `json:"event_count"`
		EventPayload Payload   `json:"event_payload"`
		ExpireAt     string    `json:"expire_at"`
		JobID        int       `json:"job_id"`
		Parameters   Params    `json:"parameters"`
		UserID       string    `json:"user_id"`
		UUID         string    `json:"uuid"`
		WorkspaceID  string    `json:"workspace_id"`
	}{
		CreatedAt: originalPayload.CreatedAt,
		CustomVal: "GW",
		EventPayload: Payload{
			Batch: originalPayload.Payload,
		},
		UUID: uuid.Must(uuid.NewRandom()).String(),
		Parameters: Params{
			SourceID: sourceId,
		},
		EventCount: 1,
		UserID:     originalPayload.UserID,
	}
```

File: enterprise/replay/replayer/transformer.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	CustomVal    string    `json:"custom_val"`
+	EventCount   int       `json:"event_count"`
+	EventPayload Payload   `json:"event_payload"`
+	ExpireAt     string    `json:"expire_at"`
+	JobID        int       `json:"job_id"`
+	Parameters   Params    `json:"parameters"`
+	UserID       string    `json:"user_id"`
+	UUID         string    `json:"uuid"`
+	WorkspaceID  string    `json:"workspace_id"`
+}
```
Comment (atzoum) [inline enterprise/replay/replayer/transformer.go:28]: These are anonymous structs at heart!
```suggestion
```

File: enterprise/replay/replayer/transformer.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	} else {
+		sourceId = strings.Split(path, "/")[0]
+	}
+	desiredPayload := DesiredPayload{
+		CreatedAt: originalPayload.CreatedAt,
+		CustomVal: "GW",
+		EventPayload: Payload{
+			Batch: originalPayload.Payload,
+		},
+		UUID: uuid.Must(uuid.NewRandom()).String(),
```
Comment (atzoum) [inline enterprise/replay/replayer/transformer.go:59]:
```suggestion
		UUID: uuid.NewString(),
```

File: enterprise/replay/replayer/transformer.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+type Payload struct {
+	Batch      json.RawMessage `json:"batch"`
+	ReceivedAt string          `json:"receivedAt"`
+	RequestIP  string          `json:"requestIP"`
+	WriteKey   string          `json:"writeKey"`
+}
+type Params struct {
+	SourceID        string `json:"source_id"`
+	SourceJobRunID  string `json:"source_job_run_id"`
+	SourceTaskRunID string `json:"source_task_run_id"`
```
Comment (atzoum) [inline enterprise/replay/replayer/transformer.go:38]: Are these needed?
```suggestion
```

**Review by atzoum — APPROVED** [review]

**Review by BonapartePC — APPROVED** [review]

**Pull request merged**